src/main/kotlin/com/hibob/academy/dao/PetDao.kt

```diff
@@ -0,0 +1,86 @@
+package com.hibob.academy.dao
+
+import org.jooq.DSLContext
+import org.springframework.stereotype.Component
+import jakarta.inject.Inject
+import org.jooq.Record
+import org.jooq.RecordMapper
+import java.time.LocalDate
+
+@Component
+class PetDao @Inject constructor(private val sql: DSLContext) {
+
+    private val petTable = PetTable.petInstance
+    private val ownerTable = OwnerTable.ownerInstance
+
+    private val petTableMapper = RecordMapper<Record, Pet> { record ->
+        Pet(
+            id = record[petTable.id],
+            name = record[petTable.name],
+            type = PetType.fromString(record[(petTable.type)]),
+            companyId = record[petTable.companyId],
+            dateOfArrival = record[petTable.dateOfArrival],
+            ownerId = record[petTable.ownerId]
+        )
+    }
+
+    fun insertPet(name: String, type: PetType, companyId: Long, dateOfArrival: LocalDate, ownerId: Long? = null) {
+        sql.insertInto(petTable)
+            .set(petTable.name, name)
+            .set(petTable.type, type.toString())
+            .set(petTable.companyId, companyId)
+            .set(petTable.dateOfArrival, dateOfArrival)
+            .set(petTable.ownerId, ownerId)
+            .execute()
+    }
+
+    fun getPetsByType(type: PetType, companyId: Long): List<Pet> {
+        return sql.select()
+            .from(petTable)
+            .where(petTable.type.eq(type.toString()))
+            .and(petTable.companyId.eq(companyId))
+            .fetch(petTableMapper)
+    }
+
+    fun getPetId(name: String, type: PetType, companyId: Long, dateOfArrival: LocalDate) :Long? {
+        return sql.select(petTable.id)
+            .from(petTable)
+            .where(petTable.name.eq(name)
+            .and(petTable.companyId.eq(companyId))
+            .and(petTable.type.eq(type.toString()))
+            .and(petTable.dateOfArrival.eq(dateOfArrival)))
+            .limit(1)
+            .fetchOne(petTable.id)
+    }
+
+    fun adopt(petId: Long, ownerId: Long, companyId: Long) {
+        sql.update(petTable)
+            .set(petTable.ownerId, ownerId)
+            .where(petTable.id.eq(petId)
+                .and(petTable.companyId.eq(companyId)))
+            .execute()
+    }
+
+    fun getPetOwnerId(petId: Long): Long? {
+        return sql.select(petTable.ownerId)
+            .from(petTable)
+            .where(petTable.id.eq(petId))
+            .fetchOne(petTable.ownerId)
+    }
+
+    fun getPetOwner(petId: Long): Owner? {
+        return sql.select()
+            .from(petTable)
+            .join(ownerTable)
```

Comment:
Do we need this join operation or a different one?
What are the differences between all the joins?

Reply:
Inner Join: Includes only rows where there is a match in both tables.
Left Join: Includes all rows from the left table (petTable), with matching rows from the right table (ownerTable). Rows from the left table with no match in the right table have NULL values for the right table’s columns.
Right Join: Includes all rows from the right table (ownerTable), with matching rows from the left table (petTable). Rows from the right table with no match in the left table have NULL values for the left table’s columns.

-> as inner row gives 1 relevant row only, and the others give more than once, I think this is the best - I get one row of pet & owner row, and from it, get the owner data. the right and left join would give more rows, which seems unnecessary, isn't it?